### PR TITLE
GH-35099: [CI][Packaging] Upgrade vcpkg to 2023.04.15 Release

### LINKED ISSUE #35099: [CI][Packaging] Pin vcpkg to release that includes OpenSSL linux-headers fix
### Describe the enhancement requested

Newer versions of vcpkg are required when building manylinux_2_28 arm64 platforms due to the following build issue with OpenSSL

- https://github.com/microsoft/vcpkg/issues/29674

since fixed in:

- https://github.com/microsoft/vcpkg/pull/29966

#34818 pins vcpkg to an untagged commit:  see https://github.com/apache/arrow/pull/34818#discussion_r1164203060, but it would be better to pin vcpkg to an official release tag.

### Component(s)

Continuous Integration, Packaging, Release

### PR DESCRIPTION
**Pull request by sjperkins**
### Rationale for this change

- https://github.com/apache/arrow/pull/34818 pinned vcpkg to a non-release master branch commit.
- This was to include unreleased changes, specifically https://github.com/microsoft/vcpkg/issues/29674.

### What changes are included in this PR?

This pins vcpkg to the official 2023.04.15 release (commit https://github.com/microsoft/vcpkg/commit/501db0f17ef6df184fcdbfbe0f87cde2313b6ab1)

### Are these changes tested?

I've successfully test that the following local wheel builds succeed:

```bash
$ ARCH=amd64 PYTHON=3.10 archery docker run python-wheel-manylinux-2014
$ ARCH=amd64 PYTHON=3.10 archery docker run python-wheel-manylinux-2-28
```

### Are there any user-facing changes?

I don't believe so

* Closes: #35099

### CODE REVIEW

File: ci/vcpkg/ports.patch [hunk 1]
```diff
@@ -49,18 +49,5 @@ index 0000000000..a57ce0c22f
...
-+++ b/scripts/cmake/vcpkg_find_acquire_program.cmake
-@@ -239,7 +239,7 @@ function(vcpkg_find_acquire_program program)
-             set(paths_to_search "${DOWNLOADS}/tools/python/${tool_subdirectory}")
-             vcpkg_list(SET post_install_command "${CMAKE_COMMAND}" -E rm python310._pth)
-         else()
--            set(program_name python3)
-+            set(program_name python)
-             set(brew_package_name "python")
-             set(apt_package_name "python3")
-         endif()
```
Comment (sjperkins) [inline ci/vcpkg/ports.patch:66]: The above patch no longer cleanly applies due to https://github.com/microsoft/vcpkg/pull/30780.

Having removed it, the following local builds succeed.

```bash
$ ARCH=amd64 PYTHON=3.10 archery docker run python-wheel-manylinux-2014
$ ARCH=amd64 PYTHON=3.10 archery docker run python-wheel-manylinux-2-28
```

However, it's probably worth running the entire suite to check if everything is OK

File: .env [hunk 2]
```diff
@@ -92,12 +92,7 @@ DEVTOOLSET_VERSION=
 # Used through docker-compose.yml and serves as the default version for the
 # ci/scripts/install_vcpkg.sh script. Prefer to use short SHAs to keep the
 # docker tags more readable.
-#
-# Please also update the crossbow configuration in order to keep the github
-# actions cache up to date for the macOS wheels:
-#   https://github.com/ursacomputing/crossbow/blob/master/.github/workflows/cache_vcpkg.yml
-# TODO: Update to an official release tag https://github.com/apache/arrow/issues/35099
```
Comment (sjperkins) [inline .env:99]: Removed due to suggestion here: https://github.com/apache/arrow/pull/34818#discussion_r1184321238

File: ci/vcpkg/ports.patch (hunk 1 above)
Comment (sjperkins) [reply ci/vcpkg/ports.patch:66]: I'll submit crossbow builds once the initial CI checks complete

**Comment by kou** [comment]
@github-actions crossbow submit  wheel-manylinux-* java-jars

File: .env (hunk 2 above)
Comment (kou) [reply .env:99]: Thanks!

File: ci/vcpkg/ports.patch (hunk 1 above)
Comment (kou) [reply ci/vcpkg/ports.patch:66]: Oh, sorry. I've done it!

Comment (sjperkins) [reply ci/vcpkg/ports.patch:66]: > Oh, sorry. I've done it!

Thank you!

**Comment by sjperkins** [comment]
All green, except for the `java-jars` which looks to be related to a maven timeout?

**Comment by kou** [comment]
I've re-ran it.

**Comment by sjperkins** [comment]
Failed again. On both occasions it seems related to transfer of the maven shade plugin?

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.1.1:shade (default) on project arrow-vector: 
Execution default of goal org.apache.maven.plugins:maven-shade-plugin:3.1.1:shade failed:
Plugin org.apache.maven.plugins:maven-shade-plugin:3.1.1 or one of its dependencies could not be resolved:
Failed to collect dependencies at org.apache.maven.plugins:maven-shade-plugin:jar:3.1.1 -> org.ow2.asm:asm:jar:6.0:
Failed to read artifact descriptor for org.ow2.asm:asm:jar:6.0:
The following artifacts could not be resolved: org.ow2.asm:asm:pom:6.0 (absent):
Could not transfer artifact org.ow2.asm:asm:pom:6.0 from/to apache (https://repo.maven.apache.org/maven2/):
Connection reset -> [Help 1]
```

**Comment by kou** [comment]
Ah, the `java-jars` nightly builds on main were failed since 2023-05-05:

https://lists.apache.org/list.html?builds@arrow.apache.org

2023-05-04: https://github.com/ursacomputing/crossbow/actions/runs/4880720274/jobs/8715652875
2023-05-05: https://github.com/ursacomputing/crossbow/actions/runs/4891509316/jobs/8732835157

So I think that the failure isn't related to this PR.

Could you open an issue for the failure? Then we can merge this.

**Review by assignUser — APPROVED** [review]
+1 Thanks!

**Comment by sjperkins** [comment]
> Could you open an issue for the failure? Then we can merge this.

https://github.com/apache/arrow/issues/35463 opened.

**Review by kou — APPROVED** [review]
+1

Thanks!

**Comment by kou** [comment]
@github-actions crossbow submit wheel-windows-*

**Comment by kou** [comment]
I want to check whether https://github.com/ursacomputing/crossbow/actions/runs/4906241298/jobs/8760550661#step:7:1397 is caused by this or not:

```text
CMake Error at scripts/cmake/z_vcpkg_apply_patches.cmake:34 (message):
  Applying patch failed: error: corrupt patch at line 18

Call Stack (most recent call first):
  scripts/cmake/vcpkg_extract_source_archive.cmake:153 (z_vcpkg_apply_patches)
  scripts/cmake/vcpkg_extract_source_archive_ex.cmake:8 (vcpkg_extract_source_archive)
  scripts/cmake/vcpkg_from_github.cmake:113 (vcpkg_extract_source_archive_ex)
  ports/snappy/portfile.cmake:4 (vcpkg_from_github)
  scripts/ports.cmake:147 (include)


error: building snappy:amd64-windows-static-md-release failed with: BUILD_FAILED
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+snappy
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?title=[snappy]+Build+error&body=Copy+issue+body+from+C%3A%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md

```

**Comment by sjperkins** [comment]
> wheel-windows-cp37-amd64

Looks like it may have. I've opened https://github.com/apache/arrow/issues/35503

**Pull request merged**